### PR TITLE
hostname: Use hostnamectl for Systemd strategy

### LINKED ISSUE #59438: hostname module systemd strategy using hostname
##### SUMMARY
On Fedora the systemd strategy is used, which is fine, but getting the current hostname uses the `hostname` command which may not always be present.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
hostname

##### ANSIBLE VERSION

```paste below
ansible 2.8.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/duck/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/duck/.local/lib/python3.7/site-packages/ansible
  executable location = /home/duck/.local/bin/ansible
  python version = 3.7.4 (default, Jul 11 2019, 10:43:21) [GCC 8.3.0]
```

##### CONFIGURATION

```paste below
ANSIBLE_FORCE_COLOR(env: ANSIBLE_FORCE_COLOR) = True
DEFAULT_VAULT_PASSWORD_FILE(env: ANSIBLE_VAULT_PASSWORD_FILE) = …
```

##### OS / ENVIRONMENT
The control node is using Debian unstable. I'm running molecule to run a test in Fedora 29.

##### STEPS TO REPRODUCE
The following is run on two test nodes, on Fedora 29 where it fails and one Centos & where it works fine.

```yaml
- name: Prepare hosts
  hosts: all
  tasks:
    - name: set proper hostname
      hostname:
        name: "{{ ansible_nodename }}.example.com"
```

##### EXPECTED RESULTS
The hostname should be set without error.

##### ACTUAL RESULTS
The systemd strategy uses hostnamectl to get and set the current and permanent hostnames except in `get_current_hostname()` where is uses `hostname` instead of `hostnamectl --transient status`. As in a container this command is not necessarily present, it fails with the following error:

```paste below
    The full traceback is:
      File "/tmp/ansible_hostname_payload_gb0hppku/ansible_hostname_payload.zip/ansible/module_utils/basic.py", line 1974, in get_bin_path
        bin_path = get_bin_path(arg, required, opt_dirs)
      File "/tmp/ansible_hostname_payload_gb0hppku/ansible_hostname_payload.zip/ansible/module_utils/common/process.py", line 41, in get_bin_path
        raise ValueError('Failed to find required executable %s in paths: %s' % (arg, os.pathsep.join(paths)))
    
    fatal: [ansible-test-builder]: FAILED! => {
        "changed": false,
        "invocation": {
            "module_args": {
                "name": "ansible-test-builder.example.com"
            }
        },
        "msg": "Failed to find required executable hostname in paths: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
    }
```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Use hostnamectl command to get current hostname for host while using
systemd strategy.

Fixes: #59438

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/system/hostname.py

### CODE REVIEW

**Comment by duck-rh** [comment]
@Akasurde I'm sorry I must miss something totally obvious to both the old code and this PR but what's the point of using the hostname command at all? What is the drawback compared to using hostnamectl unconditionally in the Systemd strategy?

**Comment by Akasurde** [comment]
> @Akasurde I'm sorry I must miss something totally obvious to both the old code and this PR but what's the point of using the hostname command at all? What is the drawback compared to using hostnamectl unconditionally in the Systemd strategy?

I kept old `hostname` command in conditional for backward compatibility. I can remove if we want `hostnamectl` unconditionally in the systemd strategy.

**Comment by duck-rh** [comment]
The documentation for hostnamectl says it uses the network configuration but it does not say which system call is used. It `seems` to return the same value as hostname, at least on my systems.

So I looked in the systemd source and in src/hostname/hostnamed.c I can see that the `Hostname` (returned when you ask for the transient hostname) is defined in PROP_HOSTNAME which comes from uname().nodename. hostname is supposed to use the gethostname() system call but according to the manpage the GlibC in fact uses uname().nodename. So if I am not mistaken, it is safe to remove the backward compatibility code as both values are always identical.

**Comment by Akasurde** [comment]
@duck-rh Could you please review this ? Thanks.

File: lib/ansible/modules/system/hostname.py [hunk 1]
```diff
@@ -374,7 +376,7 @@ class SystemdStrategy(GenericStrategy):
     """
 
     def get_current_hostname(self):
-        cmd = ['hostname']
+        cmd = ['hostnamectl', '--transient', 'status']
```
Comment (samdoran) [inline lib/ansible/modules/system/hostname.py]: We need to run `get_bin_path()` on `hostnamectl`

**Comment by samdoran** [comment]
@duck-rh The `hostname` command is needed for older systems that do not have `hostnamectl` present, e.g., RHEL/CentOS 6.

**Comment by duck-rh** [comment]
@samdoran Thanks for the explanation.

**Review by duck-rh — APPROVED** [review]
Looks good. I tested in my container where systemd is present but not hostname and it works fine now. Thanks.

File: lib/ansible/modules/system/hostname.py [hunk 2]
```diff
@@ -155,7 +155,6 @@ class GenericStrategy(object):
 
     def __init__(self, module):
         self.module = module
-        self.hostname_cmd = self.module.get_bin_path('hostname', True)
```
Comment (bcoca) [inline lib/ansible/modules/system/hostname.py:158]: keep this here, just add a attempt to get hostnamectl, then 'if present' you can use it over hostname, but fallback to current behaviour when it is not present, that way you keep backwards compatibility

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/system/hostname.py [hunk 3]
```diff
@@ -374,7 +376,7 @@ class SystemdStrategy(GenericStrategy):
     """
 
     def get_current_hostname(self):
-        cmd = ['hostname']
+        cmd = [self.hostname_cmd, '--transient', 'status']
```
Comment (bcoca) [inline lib/ansible/modules/system/hostname.py:379]: ~the options are not compatible, set `self.hostnamectl_cmd`, use that and appropriate options when populated, default to previous code when not~
nvmd, was only looking at the changes, reading the more general code, this is the systemd strategy and it always relies on hostnamectl being present, scratch what i mentioned before, just using hostnamectl in general should be fine

**Pull request merged**